File: Difficulty: Easy/Sum of elements between k1'th and k2'th smallest elements/sum-of-elements-between-k1th-and-k2th-smallest-elements.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{
    public:
  long long sumBetweenTwoKth(long long A[], long long N, long long K1, long long K2)
{
    // Your code goes here
    priority_queue<long long> maxHeap1;
    priority_queue<long long> maxHeap2;
    for (long long i = 0; i < N; i++)
    {
        maxHeap1.push(A[i]);
        if (maxHeap1.size() > K1)
            maxHeap1.pop();
        maxHeap2.push(A[i]);
        if (maxHeap2.size() > K2)
            maxHeap2.pop();
    }
     long long sum = -maxHeap2.top();
    while (!maxHeap2.empty())
    {
        sum += maxHeap2.top();
        maxHeap2.pop();
    }
    while (!maxHeap1.empty())
    {
        sum -= maxHeap1.top();
        maxHeap1.pop();
    }
    return sum;
}
};

//{ Driver Code Starts.
int main()
 {
    long long t;
    cin>>t;
    while(t--)
    {
        long long n, k;
        cin>>n;
        long long a[n+5];
        for(int i =0;i<n;i++)
            cin >> a[i];
        
        long long k1, k2;
        cin >> k1 >> k2;
        Solution ob;
        cout << ob.sumBetweenTwoKth(a, n, k1, k2) << endl;
        
    
cout << "~" << "\n";
}
	return 0;
}

// } Driver Code Ends

File: Difficulty: Easy/Sum of elements between k1'th and k2'th smallest elements/sum-of-elements-between-k1th-and-k2th-smallest-elements.java
// User function Template for Java

class Solution {
    private static long kthElement(long A[], long K) {
        PriorityQueue<Long> maxHeapk = new PriorityQueue<>((a, b) -> Long.compare(b, a));
        for (int i = 0; i < A.length; i++) {
            maxHeapk.add(A[i]);
            if (maxHeapk.size() > K)
                maxHeapk.poll();
        }
        return maxHeapk.peek();

    }

    public static long sumBetweenTwoKth(long A[], long N, long K1, long K2) {
        // Your code goes here
        long k1Smallest = kthElement(A, K1);
        long k2Smallest = kthElement(A, K2);
        long sum = 0;
        for (long i : A) {
            if (i > k1Smallest && i < k2Smallest)
                sum += i;
        }
        return sum;

    }
}
